File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-python/src/parallel_consumer/options.py
# Copyright (C) 2026 Antony Stubbs and contributors

"""Connect-time configuration: what the user tells Parallel Consumer, in one object.

Configuration is code (the language-proxy plan's KTD5): this library reads no file, no
environment variable and no command line, and nothing here reaches the sidecar by argv or
environment - it all travels in the connect-time ``Configure`` message and nowhere else.

**Unset means "take the engine's default".** This class therefore holds no defaults of its own
and never guesses one; ``None`` is the wire's absence, and the proxy reports back the value it
actually used.
"""

from __future__ import annotations

import dataclasses
import enum
from datetime import timedelta
from collections.abc import Mapping

__all__ = ["ClientOptions", "ProcessingOrder"]


class ProcessingOrder(enum.Enum):
    """The ordering guarantee to ask Parallel Consumer for.

    Ordering is enforced by the engine, never by this library.
    """

    UNORDERED = "unordered"
    """No ordering: any record may process concurrently with any other."""

    PARTITION = "partition"
    """Records of one partition process one at a time, in offset order."""

    KEY = "key"
    """Records sharing a key process one at a time; distinct keys proceed concurrently."""


@dataclasses.dataclass(frozen=True)
class ClientOptions:
    """Everything the proxy needs to build its consumer and start dispatching.

    ``kafka_properties`` is credential-bearing (bootstrap servers, SASL secrets). It is never
    logged, never echoed back in an error, and never written to argv, an environment variable or
    a temp file - which is why :meth:`__repr__` omits it. Keep that property when adding a field.

    :param topics: the subscription, fixed for the client's lifetime. Give this or
        ``topic_pattern``, never both.
    :param topic_pattern: a subscription regex, as the alternative to ``topics``.
    :param max_concurrency: the most records the proxy may have in flight to this client at
        once. ``None`` takes the proxy's default; the effective value comes back in the
        handshake and becomes this client's dispatch-queue depth.
    :param ordering: the ordering guarantee; ``None`` takes the proxy's default.
    :param commit_interval: how often offsets are committed; ``None`` takes the proxy's default.
    :param default_message_retry_delay: how long a failed record waits before redelivery.
    :param kafka_properties: the Kafka client configuration. Credential-bearing.
    """

    topics: tuple[str, ...] = ()
    topic_pattern: str | None = None
    max_concurrency: int | None = None
    ordering: ProcessingOrder | None = None
    commit_interval: timedelta | None = None
    default_message_retry_delay: timedelta | None = None
    kafka_properties: Mapping[str, str] = dataclasses.field(default_factory=dict)

    def __post_init__(self) -> None:
        # Accept any iterable of topics but hold a tuple, so an option object cannot be mutated
        # out from under a running session.
        object.__setattr__(self, "topics", tuple(self.topics))
        object.__setattr__(self, "kafka_properties", dict(self.kafka_properties))

        if bool(self.topics) == bool(self.topic_pattern):
            raise ValueError("give exactly one of topics or topic_pattern")
        if self.max_concurrency is not None and self.max_concurrency < 1:
            raise ValueError(f"max_concurrency must be at least 1, got {self.max_concurrency}")

    def __repr__(self) -> str:
        # Deliberately omits kafka_properties: it may carry credentials, and a repr reaches logs
        # and tracebacks by every route there is.
        return (
            f"ClientOptions(topics={self.topics!r}, topic_pattern={self.topic_pattern!r}, "
            f"max_concurrency={self.max_concurrency!r}, ordering={self.ordering!r}, "
            f"commit_interval={self.commit_interval!r}, "
            f"default_message_retry_delay={self.default_message_retry_delay!r})"
        )

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-typescript/src/options.ts
// Copyright (C) 2026 Antony Stubbs and contributors

import { ConfigurationError } from "./errors";

/**
 * How to launch the sidecar proxy: an absolute path to a binary, and its arguments.
 *
 * THE APPLICATION SUPPLIES THIS EXPLICITLY, and the library never goes looking. No `PATH` lookup,
 * no relative resolution, no directory an attacker could influence - this process is about to be
 * handed the application's Kafka credentials, so which binary runs is a security decision and it
 * belongs to the application (client-authoring guide §2).
 *
 * `args` are the sidecar's own arguments and are NOT configuration: configuration is code and
 * travels in the connect-time `Configure` message, never by argv, environment or file.
 */
export interface SidecarCommand {
  readonly executable: string;
  readonly args?: readonly string[];
  /**
   * Where the sidecar's stderr goes. `"inherit"` (the default) puts it on this process's stderr,
   * which is what makes a sidecar that fails at startup say so.
   */
  readonly stderr?: "inherit" | "ignore";
}

/** Processing order, spelled as a string union - the TypeScript idiom for a closed set of names. */
export type Ordering = "unordered" | "partition" | "key";

/** Commit mode. The transactional mode is expressible on the wire and always refused by the proxy. */
export type CommitMode = "periodic-consumer-sync" | "periodic-consumer-async";

/**
 * Connect-time configuration: what the application tells Parallel Consumer, in the application's
 * own language, handed over at construction.
 *
 * A PLAIN OBJECT, NOT A BUILDER. An omitted property means "take the proxy's default" - which is
 * exactly what the wire means by an absent field, so the two conventions agree with no translation
 * table in between, and `Configured` reports what each default resolved to.
 *
 * `kafkaProperties` is credential-bearing. It is written to the stream and nowhere else: never to
 * argv, never to an environment variable, never to a temp file, and never to a log line at any
 * level (client-authoring guide §6).
 */
export interface ClientOptions {
  /** How to launch the sidecar. Required - the library will not go looking for one. */
  readonly sidecar: SidecarCommand;

  /** Topic-list subscription. Exactly one of `topics` / `topicPattern` must be given. */
  readonly topics?: readonly string[];
  /** Regex subscription, in Java regex syntax - the proxy compiles it. */
  readonly topicPattern?: string;

  /**
   * The in-flight ceiling: the most records the proxy will have out to this client at once, and
   * therefore this client's dispatch-queue depth. There is no "unlimited".
   */
  readonly maxConcurrency?: number;

  /** Kafka client configuration - bootstrap servers, group id, credentials. Never logged. */
  readonly kafkaProperties?: Readonly<Record<string, string>>;

  readonly ordering?: Ordering;
  readonly commitMode?: CommitMode;

  /** Milliseconds between offset commits. */
  readonly commitIntervalMs?: number;
  /** Milliseconds a failed record waits before redelivery. */
  readonly defaultMessageRetryDelayMs?: number;

  /** Instance tag for the engine's own metrics and logging. */
  readonly instanceTag?: string;

  /**
   * Called when the client notices something it will not act on - an un-negotiated message, say.
   * Default: silence. Whatever is passed here must never be given the options object itself, which
   * would put `kafkaProperties` in a log line.
   */
  readonly onWarning?: (message: string) => void;

  /** How long to wait for the sidecar to announce its port. Default 30s. */
  readonly startupTimeoutMs?: number;
  /** How long `close()` waits for the sidecar to exit before killing it. Default 10s. */
  readonly shutdownTimeoutMs?: number;
}

/** Validated, defaulted options - the shape the rest of the client uses. */
export interface ResolvedOptions extends ClientOptions {
  readonly onWarning: (message: string) => void;
  readonly startupTimeoutMs: number;
  readonly shutdownTimeoutMs: number;
}

/**
 * Checks what can be checked before anything is spawned or connected.
 *
 * Only the subscription rule is enforced here, and only because getting it wrong costs a process
 * launch and a round trip to learn `INVALID_ARGUMENT`. Everything else is the proxy's to judge:
 * duplicating its validation here would be two rules that drift apart, and the effective values
 * come back in `Configured` anyway.
 */
export function resolveOptions(options: ClientOptions): ResolvedOptions {
  const hasTopics = options.topics !== undefined && options.topics.length > 0;
  const hasPattern = options.topicPattern !== undefined && options.topicPattern.length > 0;
  if (hasTopics === hasPattern) {
    throw new ConfigurationError(
      "exactly one of topics / topicPattern must be given - the subscription is fixed for the " +
        "proxy's lifetime and the proxy refuses both or neither",
    );
  }
  if (options.sidecar.executable.length === 0) {
    throw new ConfigurationError("sidecar.executable must be the path to the sidecar binary");
  }
  if (options.maxConcurrency !== undefined && options.maxConcurrency < 1) {
    throw new ConfigurationError(
      `maxConcurrency must be at least 1, got ${options.maxConcurrency} - there is no "unlimited"`,
    );
  }
  return {
    ...options,
    onWarning: options.onWarning ?? (() => undefined),
    startupTimeoutMs: options.startupTimeoutMs ?? 30_000,
    shutdownTimeoutMs: options.shutdownTimeoutMs ?? 10_000,
  };
}
